### ISSUE TITLE
beforeLoad function

### ISSUE THREAD

**Issue by 0xDjole** [sveltejs/kit#3992]
### Describe the problem

I am using the `load` function. It takes time to fetch data. Because it takes time, my page is hanging with no feedback to the user. We can use CSR and `onMount` to fix the issue, but then you lose benefit of SSR.

End result is poor user experience.

### Describe the proposed solution

I want to have a function `beforeLoad` that will trigger before `load` function. It will set pages default props and do initial render ( Usually it would include some sort of loading indicator  in the render until data is fetched ).

When load get's resolved, page re-renders with new props with fetched data.

That way you have a loading indicator as you usually have in CSR, but data is coming from the server.

### Importance

would make my life easier

### Additional Information

PS: I am in the first week of sveltekit and SSR so I might be missing the point here. I apologize if that is the case.

_No response_

**Comment by 0xDjole** [comment]
@mustofa-id It is, but not on initial load. Only when you navigate.

In my case I am doing oAuth2 redirect callback. When oAuth2 redirect ( from google ) sends me to my callback page that does SSR `load`, `$navigating` is false. In the end, my page hangs.

Alternative solution would be if $navigating was true not only when navigating, but directly entering the url on initial load.

**Comment by Rich-Harris** [comment]
There are a few options available to you. You can use the `navigating` store to show a progress bar (the progress is fictional, but the appearance of activity is enough to make it clear that something is happening) — example here: https://github.com/sveltejs/kit/blob/54fc2101be07cf5e9a5bd862db3122b956751b62/sites/kit.svelte.dev/src/routes/__layout.svelte#L16-L18

If you want to wait until the page has rendered before loading data, then `onMount` is the place to do it (though this is generally discouraged).

Finally, if you want to `load` the data for SSR but have instant client-side navigation, you could even do something like this:

```svelte
<script context="module">
  import { browser } from '$app/env';

  export function load({ fetch }) {
    if (browser) {
      return {};
    }

    return {
      props: {
        foo: await('/foo.json').then(r => r.json())
      }
    }
  }
</script>

<script>
  import { onMount } from 'svelte';

  export let foo;

  onMount(async () => {
    if (!foo) {
      foo = await fetch('/foo.json').then(r => r.json());
    }
  });
</script>
```

In other words you already have the primitives to implement whatever behaviour you want, so there's no real need to add a new function.

**Comment by 0xDjole** [comment]
@Rich-Harris Can you elaborate a bit more on this?

The entire idea is to never do fetch in the `onMount` while having instant client load.

Every time I RELOAD the page using the example above, `onMount` always runs without `foo` and triggers the client fetch. On server-side it also runs, but I think it gets over-written by client side...

In a nutshell I don't get when does this trigger?
` return {
      props: {
        foo: await('/foo.json').then(r => r.json())
      }
    }`
